invite cleanup: Fix minor bug, add tests, check query counts, rename functions

showell opened the pull request:
These are easy prep commits that will lead up to a slightly more intense refactoring of the invites code.

timabbott commented:
These look great except `check_incoming_email` I think should be called `validate_email_can_join_realm` -- `check_incoming_email` sounds too much like part of the incoming email gateway.

Also, what do you think about moving this bundle of functions out of `actions.py`, maybe to `zerver/lib/email_validation.py`?

showell commented:
Thanks for the quick review (despite the build flake).

I'm just gonna drop the `check_incoming_email` commit, since, as I stated in the commit message, the goal is to remove that function, which I do after a few commits (albeit not in this initial PR).  Renaming it to `validate_email_can_join_realm` temporarily would be a slight win, but there would still be a lot of name soup.

I am totally in favor of splitting out functions to a new file.  These prep commits are still mostly re-namings, but that's good to know for the upcoming stuff.  (If nothing else, it will clean up some misleadingly ugly diffs.  And it's also sorta haphazard now which code is in `actions.py` vs. `models.py`.)

showell commented:
Re-pushed and ready for review.

showell commented:
P.S. We might also be overdue to break up some of our larger test suites into smaller files.  For example, we could split out `test_invites.py` from `test_signups.py`.  I'm already in a big rabbit hole, though.

showell commented:
Ok, I pushed the rest of the series.  The first four commits should be relative no-brainers (you already reviewed them once), and the first few after that should be pretty clear wins too.

showell commented:
Oops new models function should say delivery_email instead of email.  AFK

showell commented:
I fixed the delivery_email thing. ^^^

I'm gonna try to make this less of a pitfall on a separate branch.

timabbott commented on zerver/models.py:
```diff
@@ -2144,6 +2144,18 @@ def get_user_by_delivery_email(email: str, realm: Realm) -> UserProfile:
+    it has the same security caveats.  It gets multiple
+    users and returns a QuerySet, since most callers
+    only will grab two or three fields.
+    """
```
Let's also note that it doesn't do a `select_related`, so be careful to not access foreign keys like Realm after fetching it.

showell replied:
I added some wording here, but you should probably edit it.  I think the main point to emphasize is that "most callers only will grab two or three fields", and we maybe just need to say "should only grab two..." instead of "will only grab two".

timabbott commented on zerver/lib/email_validation.py:
```diff
@@ -126,8 +125,16 @@ def get_existing_user_errors(
+        'is_active',
+        'is_mirror_dummy',
+    )
+    user_dict = {user.email: user for user in users}
```
I think this logic has a case bug -- if the user submits an email in funny case, it may have a case-insensitive match in `get_users_by_delivery_email` but miss the actual match below.

We should probably just do `.lower()` both here and in a `existing_user_profile = user_dict.get(email.lower())` below.

showell replied:
I fixed this and added a test to prevent future similar bugs.  (I pushed the test to the front before fixing the above, and it was an actual failing test till I addressed your comment, so we should be good.)

timabbott commented on zerver/models.py:
```diff
@@ -2144,6 +2144,18 @@ def get_user_by_delivery_email(email: str, realm: Realm) -> UserProfile:
+    """
+    email_filter = Q()
+    for email in emails:
+        email_filter |= Q(delivery_email__iexact=email.strip())
```
Can we just use `delivery_email__iexact__in=` and pass a list, rather than using `Q`?  That might be more readable and also produce between SQL.

showell replied:
It's a Django limitation.  Try this in the shell:

~~~
In [2]: list(UserProfile.objects.filter(email__iexact='x'))
In [3]: list(UserProfile.objects.filter(email__in=['x']))
In [4]: list(UserProfile.objects.filter(email__iexact__in=['x']))
~~~

The first two work, the third explodes.

See https://stackoverflow.com/questions/14907525/how-can-i-chain-djangos-in-and-iexact-queryset-field-lookups.

timabbott replied:
Good to know :)

timabbott commented on zerver/lib/email_validation.py:
```diff
@@ -160,20 +163,23 @@ def process_email(email: str) -> None:
         deactivated = not existing_user_profile.is_active
 
         if existing_user_profile.is_active:
```
This line would probably be better as `if not deactivated` (or better, swap the if statement and do `if deactivated`.

showell replied:
Can we add this is as a cleanup commit?  I think there's some readability in the handling the common case first, and some sort of double-negative is kinda hard to avoid here, so I'm a little wary to tweak this too much from the original code.

timabbott replied:
Yeah definitely.

showell commented:
For your naming concerns, can you address validate_email_is_valid/validate_email_is_allowed_in_realm as an additive commit.  I'm not sure I follow your logic there.   There is some shimming that goes on the intermediate commits, but here's the final version, which shows that `validate_email_is_valid` is doing more than just realm-specific checks (i.e. it also calls Django's basic email validator).

~~~ py
 93 def validate_email_is_valid(
 94     email: str,
 95     validate_email_allowed_in_realm: Callable[[str], None],
 96 ) -> Optional[str]:
 97
 98     try:
 99         validators.validate_email(email)
100     except ValidationError:
101         return _("Invalid address.")
102
103     try:
104         validate_email_allowed_in_realm(email)
105     except DomainNotAllowedForRealmError:
106         return _("Outside your domain.")
107     except DisposableEmailError:
108         return _("Please use your real email address.")
109     except EmailContainsPlusError:
110         return _("Email addresses containing + are not allowed.")
111
112     return None

showell commented:
I believed I've addressed all the comments here, and this is ready to merge.  @timabbott here are things to look for:

- tweak my comments in general, but specifically the one about `select_related`
- decide whether flipping the deactivated conditional is necessary (can you just do that as an additive commit?)
- decide whether any names can be improved (again, probably easiest to do as a final cleanup commit after the dust has settled a bit)

Other than addressing your other comments, not much has changed here.  I did split up one of the commits into two different ones.  I also added a commit to clean up the existing-user test and make it prevent case errors.

timabbott commented:
> For your naming concerns, can you address validate_email_is_valid/validate_email_is_allowed_in_realm as an additive commit? I'm not sure I follow your logic there. There is some shimming that goes on the intermediate commits, but here's the final version, which shows that validate_email_is_valid is doing more than just realm-specific checks (i.e. it also calls Django's basic email validator).

My concern is that `validate_email_is_valid` sounds more like it's JUST checking the email address format, and not realm_domains.  Maybe we just want to call that outer function `check_email_format_and_realm_domains`.  I'm open to other names.  

But the thing I feel strongest about is that code interacting with the return value of `get_realm_email_validator` should call the thing `realm_email_validator`; that piece would make things a lot easier to follow.

The related thing on my mind is that it'd be kinda nice to fold the Django `validators.validate_email` check into `email_allowed_for_realm`, such that there isn't a common way to use `email_validation.py` that involves skipping `validators.validate_email`.  (Which I think LDAP may be accidentally doing).  I'm fine with our checking that twice in the registration code path if the code flow makes that simpler to reuse code.

timabbott commented:
Merged, thanks @showell!  The rest of these things can be fixed in a follow-up PR.
